qiskit_optimization/applications/sk_model.py

```diff
@@ -0,0 +1,110 @@
+from ..translators import from_docplex_mp
+
+
+class SKModel(OptimizationApplication):
+    r"""Optimization application of the "Sherrington Kirkpatrick (SK) model" [1].
+
+    The SK Hamiltonian over n spins is given as:
+        :math:`H(x)=-1/\sqrt{n} \sum_{i<j} w_{i,j}x_ix_j`, :math:`where x_i\in\{\pm 1\}`
+    and :math:`w_{i,j}\in\{\pm 1\}` are chosen independently and uniformly at random.
+    Notice that there are other variants e.g., with :math:`w_{i,j}` chosen from the normal distribution
+    with mean 0 and variance 1.
+
+    References:
+        [1]: Dmitry Panchenko. "The Sherrington-Kirkpatrick model: an overview",
+        https://arxiv.org/abs/1211.1094
+    """
+
+    def __init__(self, num_of_sites: int, rng: np.random.RandomState = np.random.RandomState(0)):
+        """
+        Constructor for the SK model.
+
+        Args:
+            num_of_sites: number of sites
+            rng: numpy pseudo-random number generator
+        """
+        self._rng = rng
+        self._num_of_sites = num_of_sites
+        self._graph = nx.complete_graph(self._num_of_sites)
+
+        self.new_instance()
+
+    def new_instance(self) -> None:
+        """Generate a new instance of the SK model."""
+        for _, _, edge_data in self._graph.edges(data=True):
+            edge_data["weight"] = self._rng.choice([-1, 1])
+
+    def to_quadratic_program(self) -> QuadraticProgram:
+        """Convert an SK model problem instance into a
+        :class:`~qiskit_optimization.problems.QuadraticProgram`.
+
+        Returns:
+            The :class:`~qiskit_optimization.problems.QuadraticProgram` created
+            from the SK problem instance.
+        """
+        mdl = Model(name="SK-model")
+        x = {
+            i: mdl.binary_var(name="x_{0}".format(i)) for i in range(self._graph.number_of_nodes())
+        }
+
+        objective = mdl.sum(
+            -1
+            / np.sqrt(self._num_of_sites)
+            * self._graph.edges[i, j]["weight"]
+            * (2 * x[i] - 1)
+            * (2 * x[j] - 1)
+            for i, j in self._graph.edges
+        )
+        # we converted the standard H(x)=-1/\sqrt{n} \sum w_{ij}x_ix_j, where x_i\in\pm 1 to binary.
+
+        mdl.minimize(objective)
+        return from_docplex_mp(mdl)
+
+    def interpret(self, result: Union[OptimizationResult, np.ndarray]) -> List[int]:
+        """Interpret a result as configuration of spins.
+
+        Args:
+            result : The calculated result of the problem.
+
+        Returns:
+            configuration of spins
+        """
+        configuration = list(map(lambda x: 2 * x - 1, self._result_to_x(result)))
+        return configuration
+
+    @property
+    def graph(self) -> nx.Graph:
+        """Getter of the graph.
+        Returns:
+            A graph for a problem.
+        """
+        return self._graph
+
+    def _set_rng(self, rng):
+        self._rng = rng
+
+    rng = property(fset=_set_rng)
```

Comment:
If you want to implement a setter, could you use the following idiom with a docstring as in the python doc?
https://docs.python.org/3/library/functions.html?highlight=setter#property
```
    @x.setter
    def x(self, value):
        """ Documentation comes here
        """
        self._x = value
```

Reply:
I am not sure why we would want an rng setter - isn't setting/influencing it via constructor enough? What does this even do since the constructor builds using the rng and I do not see it used afterwards that a setter would have any value.

Reply:
I thought it might be useful to be able to change the pseudorandom number generator.
It will be used in the next call of `new_instance()` which generates a new disorder of the SK model.
Obviously you can create a new instance of the `SKModel` class and start with a new seed.
And maybe it is not that useful after all.

Reply:
I think it's confusing to provide rng setter. If users want to change rng, they can create a new object with a new rng generator.

Reply:
ok. I take it out.